pkg/cloud/services/networking/router.go

```diff
@@ -239,19 +240,56 @@ func (s *Service) getRouterInterfaces(routerID string) ([]ports.Port, error) {
 	return portList, nil
 }
 
-func (s *Service) existsRouter(routerID string) (bool, error) {
-	if routerID == "" {
-		return false, nil
+func (s *Service) getRouterByName(routerName string) (routers.Router, error) {
+	allPages, err := routers.List(s.client, routers.ListOpts{
+		Name: routerName,
+	}).AllPages()
+	if err != nil {
+		return routers.Router{}, err
 	}
-	router, err := routers.Get(s.client, routerID).Extract()
+
+	routerList, err := routers.ExtractRouters(allPages)
 	if err != nil {
-		if errors.IsNotFound(err) {
-			return false, nil
-		}
-		return false, err
+		return routers.Router{}, err
+	}
+
+	if len(routerList) > 1 {
+		return routers.Router{}, fmt.Errorf("found %d router with the name %s, which should not happen", len(routerList), routerName)
+	}
+
+	switch len(routerList) {
+	case 0:
+		return routers.Router{}, nil
+	case 1:
+		return routerList[0], nil
+	}
+	return routers.Router{}, errors.New("too many resources")
+}
+
+func (s *Service) getSubnetByName(subnetName string) (subnets.Subnet, error) {
+	opts := subnets.ListOpts{
+		Name: subnetName,
+	}
+
+	allPages, err := subnets.List(s.client, opts).AllPages()
+	if err != nil {
+		return subnets.Subnet{}, err
+	}
+
+	subnetList, err := subnets.ExtractSubnets(allPages)
+	if err != nil {
+		return subnets.Subnet{}, err
 	}
-	if router == nil {
-		return false, nil
+
+	switch len(subnetList) {
+	case 0:
+		return subnets.Subnet{}, nil
+	case 1:
+		return subnetList[0], nil
 	}
-	return true, nil
+	return subnets.Subnet{}, errors.New("too many resources")
```

Comment:
can we we write something about routers instead of the generic resources? not sure if we have similar errors already which we should stay consistent to

Reply:
Introduced Router type.

Reply:
I think I meant it differently. I added new suggestions.